akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala

```diff
@@ -151,8 +151,16 @@ private[http] object HttpServerBluePrint {
         // and then toggle back to "idle"
 
         val entitySource = new SubSourceOutlet[RequestOutput]("EntitySource")
-        // optimization: re-use the idle outHandler
-        entitySource.setHandler(this)
+
+        entitySource.setHandler(new OutHandler {
+          override def onPull(): Unit = pull(in)
+          override def onDownstreamFinish(): Unit = {
+            // application layer has cancelled or only partially consumed response entity:
+            // connection will be closed
+            entitySource.complete()
+            completeStage()
+          }
+        })
 
         setHandler(in, new InHandler {
```

Comment:
btw this handle can also be combined with the `OutHandler` which is being set on `out` port below too.